shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -0,0 +1,524 @@
+ *
+ * {@code
+ *    String uri = FlutterJNI.nativeGetObservatoryUri();
+ * }
+ */
+public class FlutterJNI {
+  private static final String TAG = "FlutterJNI";
+
+  @UiThread
+  public static native boolean nativeGetIsSoftwareRenderingEnabled();
+
+  @UiThread
+  public static native String nativeGetObservatoryUri();
+  
+  private Long nativePlatformViewId;
+  private FlutterRenderer.RenderSurface renderSurface;
+  private PlatformMessageHandler platformMessageHandler;
+  private final Set<EngineLifecycleListener> engineLifecycleListeners = new HashSet<>();
+  private final Set<OnFirstFrameRenderedListener> firstFrameListeners = new HashSet<>();
+
+  /**
+   * Sets the {@link FlutterRenderer.RenderSurface} delegate for the attached Flutter context.
+   *
+   * Flutter expects a user interface to exist on the platform side (Android), and that interface
+   * is expected to offer some capabilities that Flutter depends upon. The {@link FlutterRenderer.RenderSurface}
+   * interface represents those expectations. For example, Flutter expects to be able to request
+   * that its user interface "update custom accessibility actions" and therefore the delegate interface
+   * declares a corresponding method, {@link FlutterRenderer.RenderSurface#updateCustomAccessibilityActions(ByteBuffer, String[])}.
+   *
+   * If an app includes a user interface that renders a Flutter UI then a {@link FlutterRenderer.RenderSurface}
+   * should be set (this is the typical Flutter scenario). If no UI is being rendered, such as a
+   * Flutter app that is running Dart code in the background, then no registration may be necessary.
+   *
+   * If no {@link FlutterRenderer.RenderSurface} is registered then related messages coming from
+   * Flutter will be dropped (ignored).
+   */
+  @UiThread
+  public void setRenderSurface(@Nullable FlutterRenderer.RenderSurface renderSurface) {
+    this.renderSurface = renderSurface;
+  }
+
+  /**
+   * Call invoked by native to be forwarded to an {@link io.flutter.view.AccessibilityBridge}.
+   *
+   * The {@code buffer} and {@code strings} form a communication protocol that is implemented here:
+   * https://github.com/flutter/engine/blob/master/shell/platform/android/platform_view_android.cc#L207
+   */
+  @SuppressWarnings("unused")
+  @UiThread
+  private void updateSemantics(ByteBuffer buffer, String[] strings) {
+    if (renderSurface != null) {
+      renderSurface.updateSemantics(buffer, strings);
+    }
+    // TODO(mattcarroll): log dropped messages when in debug mode (https://github.com/flutter/flutter/issues/25391)
+  }
+
+  /**
+   * Call invoked by native to be forwarded to an {@link io.flutter.view.AccessibilityBridge}.
+   *
+   * The {@code buffer} and {@code strings} form a communication protocol that is implemented here:
+   * https://github.com/flutter/engine/blob/master/shell/platform/android/platform_view_android.cc#L207
+   */
+  @SuppressWarnings("unused")
```

Comment:
It might be useful to describe what these are used for and maybe link to:
https://docs.flutter.io/flutter/semantics/CustomSemanticsAction-class.html

The TL;DR is that Android (and iOS) define a hardcoded enum of actions like 'tap' or 'hover' but lots of apps want to attach a list of additional actions to a given node (imagine gMail adding archive, delete, mark unread actions for a given mail -- similar to the swipe actions in the UI).

Reply:
I'll add a TODO for you since these comments don't play a significant role in the current work.